### PR TITLE
[tf.data] Implement SkipInternal for TFRecordDataset

### PR DESCRIPTION
**Pull request by zhuzilin**
This is a PR from JIZHI Team & TaiJi AI platform in Tencent.

This PR is a follow-up of #40963 and #41222. In those 2 PRs, I added the `Skip` method to `IteratorBase` and `RecordReader` so that we can potentially reduce the IO when user is using `tf.data.Dataset.shard()` method on samples instead of filenames.

This PR adds the `SkipInternal` for `TFRecordDatasetOp`. The implementation is very similar to `GetNextInternal`. And to test the new method, I added the helper functions and macros for `Skip`, which are also very similar to that of `GetNext`.

I'm not sure the proper way to deal with `metrics::GetTFDataBytesReadCounter` in `SkipInternal`, so it is omitted for now. It would be great if you could give me some hints on it.

Sorry for this late follow-up, I would try to add the meaningful `SkipInternal`s in the coming months.

cc @aaudiber @jsimsa

### CODE REVIEW

File: tensorflow/core/kernels/data/tf_record_dataset_op.cc [hunk 1]
```diff
@@ -158,6 +158,49 @@ class TFRecordDatasetOp::Dataset : public DatasetBase {
...
+    Status SkipInternal(IteratorContext* ctx, int num_to_skip,
+                        bool* end_of_sequence,
+                        int* num_skipped) override {
+      *num_skipped = 0;
+      mutex_lock l(mu_);
+      do {
+        // We are currently processing a file, so try to skip reading
+        // the next (num_to_skip - *num_skipped) record.
+        if (reader_) {
+          int num_skipped_once;
```
Comment (aaudiber) [inline tensorflow/core/kernels/data/tf_record_dataset_op.cc:170]: nit: rename to `latest_num_skipped` or `last_num_skipped`.

**Review by aaudiber — CHANGES_REQUESTED** [review]
Thanks @zhuzilin! The current behavior of not changing `GetTFDataBytesReadCounter` during `Skip` is correct. Out of curiosity, have you been able to measure speedup on real examples using `shard` with these changes?

**Comment by zhuzilin** [comment]
@aaudiber I've experimented with imagenet dataset. The test code I used is:
```python
from time import time
from tqdm import tqdm

import tensorflow as tf
from tensorflow.python.platform import gfile
# Use the internal class to bypass the flat_map
from tensorflow.python.data.ops.readers import _TFRecordDataset

glob_pattern = "~/imagenet/train-*-of-*"
file_names = gfile.Glob(glob_pattern)

num_shard = 8
ds = _TFRecordDataset(file_names)
# use "ds = ds.repeat().shard(num_shard, 0)" as the version without using skip
ds = ds.shard(num_shard, 0).repeat()

i = 0
_ = next(iter(ds))
start = time()
for _ in tqdm(ds):
    if i == 10000:
        break
    i += 1
print(time() - start)
```
The result is that when dataset is cached by OS (e.g. running the above script for the second time), the version with skip will have obvious improvement:
|num_shard|origin (iter/s)|with skip (iter/s)|
|-|-|-|
|2|6231|7925|
|4|3883|5901|
|8|2253|3949|
|16|1169|2537|

with the dataset's own speed at 9548 iter/s.

But when the dataset is not cached by OS, the performance of the two come close.

I think this is because tfrecord always needs to call `file_->Read` for the header underneath, no matter we are using `SkipRecords` or `ReadRecord`. And when file is not cached, the read operation itself is really slow, regardless of the amount of data we read.

The `shard` method after `TFRecordDataset` instead of after `list_files` is for user with one-file or uneven-size-files dataset. I think that the current speed up will give some help and I will continue to investigate why the performance is not improved for the uncached situation.

**Comment by zhuzilin** [comment]
@aaudiber Could you have another look at this pr? Thank you!

**Review by aaudiber — APPROVED** [review]

**Comment by zhuzilin** [comment]
@aaudiber Could you help me to check what is the build error? It's an internal one and I have no access to the log. Thank you~

File: tensorflow/core/kernels/data/dataset_test_base.cc [hunk 2]
```diff
@@ -607,6 +607,27 @@ Status DatasetOpsTestBase::CheckIteratorGetNext(
...
+Status DatasetOpsTestBase::CheckIteratorSkip(
+    int num_to_skip, int expected_num_skipped,
+    bool get_next, const std::vector<Tensor>& expected_outputs,
+    bool compare_order) {
+  IteratorBase* iterator = iterator_.get();
+  IteratorContext* ctx = iterator_ctx_.get();
+
+  bool end_of_sequence = false;
+  int num_skipped = 0;
+  iterator->Skip(ctx, num_to_skip, &end_of_sequence, &num_skipped);
```
Comment (aaudiber) [inline tensorflow/core/kernels/data/dataset_test_base.cc:619]: Wrap in `TF_RETURN_IF_ERROR`

**Review by aaudiber — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/data/dataset_test_base.cc (hunk 2 above)
Comment (zhuzilin) [reply tensorflow/core/kernels/data/dataset_test_base.cc]: @aaudiber Updated.

**Review by aaudiber — APPROVED** [review]

**Comment by zhuzilin** [comment]
@aaudiber There is still error in "feedback/copybara"... Would you mind checking that again? Thank you.

**Comment by aaudiber** [comment]
Here is the error:

In file included from third_party/tensorflow/core/kernels/data/tf_record_dataset_op_test.cc:14:
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:394:23: error: field 'compare_order' is uninitialized when used here [-Werror,-Wuninitialized]
        compare_order(compare_order) {}
                      ^
third_party/tensorflow/core/kernels/data/tf_record_dataset_op_test.cc:158:11: note: in instantiation of member function 'tensorflow::data::SkipTestCase<tensorflow::data::(anonymous namespace)::TFRecordDatasetParams>::SkipTestCase' requested here
  return {{/*dataset_params=*/TFRecordDatasetParams1(),

**Review by aaudiber — APPROVED** [review]

**Comment by zhuzilin** [comment]
:joy: Thank you for your fast reply and aproval. I've fixed the typo. Don't understand why other tests could pass though...

**Comment by zhuzilin** [comment]
@aaudiber 
Some update for the performance. If we set the `buffer_size` of the `_TFRecordDataset` to `0` so that the `TFRecordDatasetOp` will no longer use the underlaying `BufferedInputStream`, this PR will make `shard` have almost no performance drop when the data was cached.

num_shard | origin (iter/s) | with skip (iter/s) | skip + buffer_size=0 (iter/s)
-- | -- | -- | --
2 | 6231 | 7925 | 9963
4 | 3883 | 5901 | 9833
8 | 2253 | 3949 | 9490
16 | 1169 | 2537 | 8563

**Pull request merged**